File: src-python/core/services/undo_chapter.py
"""撤销最新章流程。参见 PRD §6.3 步骤 0-10。

⚠️ 这是整个项目最危险的 Service：10 步级联回滚，涉及 5 类文件。
任何一步遗漏都可能导致数据不一致。

严格遵循多文件写入顺序：数据变更 → state → ops（事务提交标记）。
AU 互斥锁在入口获取（D-0009）。方法是同步的（D-0021）。
"""

from __future__ import annotations

from pathlib import Path
from typing import Any, Optional

from core.domain.character_scanner import scan_characters_in_chapter
from core.domain.enums import FactStatus, IndexStatus
from core.domain.ops_entry import OpsEntry
from core.domain.text_utils import extract_last_scene_ending
from core.services.au_mutex import AUMutexManager
from infra.storage_local.file_utils import now_utc
from repositories.implementations.local_file_ops import generate_op_id
from repositories.interfaces.chapter_repository import ChapterRepository
from repositories.interfaces.draft_repository import DraftRepository
from repositories.interfaces.fact_repository import FactRepository
from repositories.interfaces.ops_repository import OpsRepository
from repositories.interfaces.state_repository import StateRepository

class UndoChapterError(Exception):
    """撤销章节流程错误。"""


class UndoChapterService:
    """撤销最新章 Service（PRD §6.3 步骤 0-10）。

    依赖通过构造器注入。方法是同步的（D-0021），API 层负责 async 包装。
    """

    def __init__(
        self,
        chapter_repo: ChapterRepository,
        draft_repo: DraftRepository,
        state_repo: StateRepository,
        ops_repo: OpsRepository,
        fact_repo: FactRepository,
        au_mutex: AUMutexManager,
        task_queue: Optional[Any] = None,
    ) -> None:
        self._chapter_repo = chapter_repo
        self._draft_repo = draft_repo
        self._state_repo = state_repo
        self._ops_repo = ops_repo
        self._fact_repo = fact_repo
        self._mutex = au_mutex
        self._task_queue = task_queue

    def undo_latest_chapter(
        self,
        au_path: Path,
        cast_registry: Optional[dict[str, Any]] = None,
        character_aliases: Optional[dict[str, list[str]]] = None,
    ) -> dict[str, Any]:
        """撤销最新章。严格遵循 PRD §6.3 步骤 0-10。

        Args:
            au_path: AU 根目录。
            cast_registry: project.yaml 的 cast_registry（快照真空兜底时需要）。
            character_aliases: {主名: [别名列表]}（快照真空兜底时需要）。

        Returns:
            撤销结果 dict（chapter_num, new_current_chapter）。

        Raises:
            UndoChapterError: 没有章节可撤销或其他前置校验失败。
        """
        au_id = str(au_path)

        with self._mutex.get_lock(au_id):
            return self._do_undo(
                au_id,
                cast_registry or {},
                character_aliases or {},
            )

    def _do_undo(
        self,
        au_id: str,
        cast_registry: dict[str, Any],
        character_aliases: dict[str, list[str]],
    ) -> dict[str, Any]:
        """锁内执行撤销流程（步骤 0-10）。"""

        # =================================================================
        # 步骤 0：前置校验 + 异步任务取消
        # =================================================================
        state = self._state_repo.get(au_id)
        n = state.current_chapter - 1

        if n < 1:
            raise UndoChapterError("没有已确认章节可撤销（current_chapter == 1）")

        # 取消该章的 vectorize_chapter 排队任务（防止幽灵数据）
        # 注：已执行的任务无法取消，但 vectorize 有防御性校验（章节不存在则跳过）

        # =================================================================
        # 步骤 1：确定被撤销的章节号 N
        # =================================================================
        # N = current_chapter - 1（D-0001）

        # 读取被撤销章节信息（用于 ops target_id）
        chapter_id = ""
        try:
            old_chapter = self._chapter_repo.get(au_id, n)
            chapter_id = old_chapter.chapter_id
        except FileNotFoundError:
            pass  # 章节文件已不存在（异常状态），继续回滚

        # =================================================================
        # 步骤 3a：facts resolves 关系回滚（在删除前执行）
        # =================================================================
        self._rollback_fact_statuses(au_id, n)

        # =================================================================
        # 步骤 3b：回放 update_fact_status 操作（PRD §6.3 步骤 3）
        # 用户在写第 N 章时手动修改的旧 fact 状态须恢复为操作前状态
        # =================================================================
        self._rollback_manual_status_changes(au_id, n)

        # =================================================================
        # 步骤 2：删除章节文件 + 清理 ≥N 的所有草稿（D-0016）
        # =================================================================
        self._chapter_repo.delete(au_id, n)
        self._draft_repo.delete_from_chapter(au_id, n)

        # =================================================================
        # 步骤 4：facts 物理删除（D-0003，通过 ops target_id 精准删除）
        # =================================================================
        self._delete_chapter_facts(au_id, n)

        # =================================================================
        # 步骤 5：ChromaDB chunks 删除（Phase 1: 标记 stale）
        # =================================================================
        if self._task_queue is not None:
            self._task_queue.enqueue("delete_chapter_chunks", au_id, {
                "au_path": au_id, "chapter_num": n,
            })
        state.index_status = IndexStatus.STALE

        # =================================================================
        # 步骤 6：last_scene_ending 回滚
        # =================================================================
        state.last_scene_ending = self._rollback_last_scene_ending(au_id, n)

        # =================================================================
        # 步骤 7：characters_last_seen 回滚
        # =================================================================
        state.characters_last_seen = self._rollback_characters_last_seen(
            au_id, n, cast_registry, character_aliases
        )

        # =================================================================
        # 步骤 8：chapter_focus 清空
        # =================================================================
        state.chapter_focus = []

        # =================================================================
        # 步骤 9：last_confirmed_chapter_focus 回退
        # =================================================================
        state.last_confirmed_chapter_focus = self._rollback_confirmed_focus(
            au_id, n
        )

        # =================================================================
        # 步骤 10：chapters_dirty 清理 + chapter_titles 清理
        # =================================================================
        if n in state.chapters_dirty:
            state.chapters_dirty.remove(n)
        state.chapter_titles.pop(n, None)

        # =================================================================
        # 更新 current_chapter
        # =================================================================
        state.current_chapter = n  # 回退：下一章待写变回 N

        # =================================================================
        # 最终写入（遵循多文件写入顺序）
        # =================================================================

        # state.yaml（StateRepository.save 自动 revision+1 + updated_at）
        self._state_repo.save(state)

        # ops.jsonl（事务提交标记，payload 为空 {}）
        ops_entry = OpsEntry(
            op_id=generate_op_id(),
            op_type="undo_chapter",
            target_id=chapter_id,
            chapter_num=n,
            timestamp=now_utc(),
            payload={},
        )
        self._ops_repo.append(au_id, ops_entry)

        return {
            "chapter_num": n,
            "new_current_chapter": state.current_chapter,
        }

    # -----------------------------------------------------------------
    # 步骤 3：facts 状态回滚
    # -----------------------------------------------------------------

    def _rollback_fact_statuses(self, au_id: str, n: int) -> None:
        """回滚被撤销章节 facts 的 resolves 关系。

        通过 ops target_id 定位第 N 章的 facts（不依赖可变的 fact.chapter 字段），
        然后对有 resolves 字段的条目恢复被指向目标的 status。
        """
        # 通过 ops 精准定位第 N 章的 facts（与步骤 4 用同一数据源）
        add_fact_ops = self._ops_repo.get_add_facts_for_chapter(au_id, n)
        ids_to_delete = {op.target_id for op in add_fact_ops}

        if not ids_to_delete:
            return

        all_facts = self._fact_repo.list_all(au_id)

        # 在即将被删除的 facts 中找有 resolves 关系的
        targets_to_check: set[str] = set()
        for fact in all_facts:
            if fact.id in ids_to_delete and fact.resolves:
                targets_to_check.add(fact.resolves)

        if not targets_to_check:
            return

        for target_id in targets_to_check:
            target = self._fact_repo.get(au_id, target_id)
            if target is None or target.status != FactStatus.RESOLVED:
                continue

            # 检查是否有其他 fact（排除即将删除的）仍然 resolves 该目标
            still_resolved = any(
                f.resolves == target_id
                for f in all_facts
                if f.id not in ids_to_delete and f.id != target_id
            )
            if not still_resolved:
                target.status = FactStatus.UNRESOLVED
                self._fact_repo.update(au_id, target)

    # -----------------------------------------------------------------
    # 步骤 3b：回放 update_fact_status 操作
    # -----------------------------------------------------------------

    def _rollback_manual_status_changes(self, au_id: str, n: int) -> None:
        """回放第 N 章中用户手动执行的 update_fact_status 操作（PRD §6.3 步骤 3）。

        扫描 ops.jsonl 中 chapter_num==N 且 op_type=="update_fact_status" 的记录，
        按逆序回放——将每条 fact 恢复为 payload.old_status。
        """
        all_ops = self._ops_repo.list_all(au_id)
        status_ops = [
            op for op in all_ops
            if op.op_type == "update_fact_status" and op.chapter_num == n
        ]

        if not status_ops:
            return

        # 按时间戳逆序回放
        status_ops.sort(key=lambda op: op.timestamp, reverse=True)

        for op in status_ops:
            old_status = op.payload.get("old_status")
            if not old_status:
                continue

            fact = self._fact_repo.get(au_id, op.target_id)
            if fact is None:
                continue

            fact.status = FactStatus(old_status)
            self._fact_repo.update(au_id, fact)

    # -----------------------------------------------------------------
    # 步骤 4：facts 物理删除
    # -----------------------------------------------------------------

    def _delete_chapter_facts(self, au_id: str, n: int) -> None:
        """通过 ops target_id 精准删除第 N 章的 facts（D-0003）。

        ⚠️ 禁止按 chapter 字段删除——chapter 是用户可变字段。
        若 ops 中无 add_fact 记录，不删除任何 fact。
        """
        add_fact_ops = self._ops_repo.get_add_facts_for_chapter(au_id, n)
        if not add_fact_ops:
            return  # 无记录则不删除（宁可残留也不误删）

        target_ids = [op.target_id for op in add_fact_ops]
        self._fact_repo.delete_by_ids(au_id, target_ids)

    # -----------------------------------------------------------------
    # 步骤 6：last_scene_ending 回滚
    # -----------------------------------------------------------------

    def _rollback_last_scene_ending(self, au_id: str, n: int) -> str:
        """回滚 last_scene_ending。

        N==1: 空字符串。
        否则: 优先从 ops 快照恢复，降级为读取 ch{N-1} 末尾。
        """
        if n == 1:
            return ""

        # 优先：ops 快照
        confirm_op = self._ops_repo.get_confirm_for_chapter(au_id, n - 1)
        if confirm_op and "last_scene_ending_snapshot" in confirm_op.payload:
            snapshot = confirm_op.payload["last_scene_ending_snapshot"]
            if isinstance(snapshot, str):
                return snapshot

        # 降级：读取 ch{N-1} 末尾
        try:
            content = self._chapter_repo.get_content_only(au_id, n - 1)
            return extract_last_scene_ending(content)
        except FileNotFoundError:
            return ""

    # -----------------------------------------------------------------
    # 步骤 7：characters_last_seen 回滚
    # -----------------------------------------------------------------

    def _rollback_characters_last_seen(
        self,
        au_id: str,
        n: int,
        cast_registry: dict[str, Any],
        character_aliases: dict[str, list[str]],
    ) -> dict[str, int]:
        """回滚 characters_last_seen。

        N==1: 空字典。
        否则: 优先从 ops 快照恢复，降级为全量扫描重建。
        """
        if n == 1:
            return {}

        # 优先：ops 快照
        confirm_op = self._ops_repo.get_confirm_for_chapter(au_id, n - 1)
        if confirm_op and "characters_last_seen_snapshot" in confirm_op.payload:
            snapshot = confirm_op.payload["characters_last_seen_snapshot"]
            if isinstance(snapshot, dict):
                try:
                    return {str(k): int(v) for k, v in snapshot.items()}
                except (ValueError, TypeError):
                    pass  # 快照数据损坏，降级到全量扫描

        # 降级：全量扫描重建（快照真空兜底）
        return self._rebuild_characters_last_seen(
            au_id, cast_registry, character_aliases
        )

    def _rebuild_characters_last_seen(
        self,
        au_id: str,
        cast_registry: dict[str, Any],
        character_aliases: dict[str, list[str]],
    ) -> dict[str, int]:
        """全量扫描所有现存章节重建 characters_last_seen。

        撤销是低频高危操作，必须优先保证数据完整性而非速度。
        """
        chapters = self._chapter_repo.list_main(au_id)
        result: dict[str, int] = {}
        for ch in chapters:
            scanned = scan_characters_in_chapter(
                ch.content, cast_registry, character_aliases, ch.chapter_num
            )
            for name, num in scanned.items():
                if num > result.get(name, 0):
                    result[name] = num
        return result

    # -----------------------------------------------------------------
    # 步骤 9：last_confirmed_chapter_focus 回退
    # -----------------------------------------------------------------

    def _rollback_confirmed_focus(self, au_id: str, n: int) -> list[str]:
        """回退 last_confirmed_chapter_focus。

        读取 ch{N-1} 的 frontmatter confirmed_focus。
        N==1 或章节不存在时清空。
        """
        if n <= 1:
            return []
        try:
            prev_ch = self._chapter_repo.get(au_id, n - 1)
            return list(prev_ch.confirmed_focus)
        except FileNotFoundError:
            return []

File: src-engine/services/undo_chapter.ts
// Copyright (c) 2026 FicForge Contributors
// Licensed under the GNU Affero General Public License v3.0.

/**
 * 撤销最新章流程。参见 PRD §6.3 步骤 0-10。
 *
 * ⚠️ 全代码库最危险的 Service：10 步级联回滚，涉及 5 类文件。
 * 使用 WriteTransaction 保证 D-0036 写入顺序：读取+计算 → tx(ops → chapters → facts → drafts → state)。
 */

import { mergeCharactersLastSeen, scanCharactersInChapter } from "../domain/character_scanner.js";
import { FactStatus, IndexStatus } from "../domain/enums.js";
import { logCatch } from "../logger/index.js";
import type { OpsEntry } from "../domain/ops_entry.js";
import { createOpsEntry } from "../domain/ops_entry.js";
import type { Fact } from "../domain/fact.js";
import { extractLastSceneEnding } from "../domain/text_utils.js";
import type { ChapterRepository } from "../repositories/interfaces/chapter.js";
import type { DraftRepository } from "../repositories/interfaces/draft.js";
import type { FactRepository } from "../repositories/interfaces/fact.js";
import type { OpsRepository } from "../repositories/interfaces/ops.js";
import type { StateRepository } from "../repositories/interfaces/state.js";
import { generateOpId, nowUtc } from "../utils/file_utils.js";
import { withAuLock } from "./au_lock.js";
import { WriteTransaction } from "./write_transaction.js";

export class UndoChapterError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "UndoChapterError";
  }
}

/**
 * undo 流程自身记账产生的 `update_fact_status` op 的 reason —— 生产端与过滤端共用，
 * 避免「字面量散落 + 前缀约定」两处独立维护漂移（单一真相源）。
 * 这些 op 不代表本章「真实发生过」的状态变更，故被 isUndoGeneratedStatusOp 排除回放。
 */
const UNDO_REASON_PREFIX = "undo_";
const UNDO_RESOLVES_CASCADE_REASON = `${UNDO_REASON_PREFIX}resolves_cascade`;
const UNDO_MANUAL_ROLLBACK_REASON = `${UNDO_REASON_PREFIX}manual_rollback`;

export interface UndoChapterParams {
  au_id: string;
  cast_registry?: { characters?: string[] };
  character_aliases?: Record<string, string[]> | null;
  chapter_repo: ChapterRepository;
  draft_repo: DraftRepository;
  state_repo: StateRepository;
  ops_repo: OpsRepository;
  fact_repo: FactRepository;
}

export interface UndoChapterResult {
  chapter_num: number;
  new_current_chapter: number;
}

/**
 * 撤销最新章入口。持 AU 锁覆盖 10 步级联回滚的整个事务 ——
 * 这是全代码库最危险的 Service，任何并发插入都会破坏级联一致性。
 * 锁分层策略见 services/au_lock.ts。
 */
export async function undoLatestChapter(params: UndoChapterParams): Promise<UndoChapterResult> {
  return withAuLock(params.au_id, () => doUndo(params));
}

async function doUndo(params: UndoChapterParams): Promise<UndoChapterResult> {
  const {
    au_id,
    cast_registry = { characters: [] },
    character_aliases = null,
    chapter_repo,
    draft_repo,
    state_repo,
    ops_repo,
    fact_repo,
  } = params;

  // =================================================================
  // 步骤 0：前置校验
  // =================================================================
  const state = await state_repo.get(au_id);
  const n = state.current_chapter - 1;

  if (n < 1) {
    throw new UndoChapterError("没有已确认章节可撤销（current_chapter == 1）");
  }

  // =================================================================
  // 步骤 1：确定被撤销的章节号 N
  // =================================================================
  let chapterId = "";
  // 章节文件已不存在（get 返回 null，异常状态）时继续回滚；fs 读错误照抛中止 undo
  // —— 旧的裸 catch 会把真实读错误也吞成"不存在"，读失败时不该带病级联。
  const oldChapter = await chapter_repo.get(au_id, n);
  if (oldChapter) chapterId = oldChapter.chapter_id;

  // =================================================================
  // 读取阶段：收集所有待写入操作（不实际写入）
  // =================================================================
  const tx = new WriteTransaction();

  // 步骤 3a：facts resolves 关系回滚
  const resolvesOps = await collectResolvesRollback(au_id, n, ops_repo, fact_repo);
  for (const { op, fact } of resolvesOps) {
    tx.appendOp(au_id, op);
    tx.updateFact(au_id, fact);
  }

  // 步骤 3b：回放 update_fact_status（收集待更新 facts + 反向 op，TD-003）
  const manualRollbacks = await collectManualStatusRollback(au_id, n, ops_repo, fact_repo);
  for (const { op, fact } of manualRollbacks) {
    tx.appendOp(au_id, op);
    tx.updateFact(au_id, fact);
  }

  // 步骤 2：删除章节文件 + 清理 ≥N 的所有草稿（D-0016）
  tx.deleteChapter(au_id, n);
  tx.deleteDraftFromChapter(au_id, n);

  // 步骤 4：facts 物理删除（D-0003，通过 ops target_id 精准删除）
  const { deleteOps, factIdsToDelete } = await collectChapterFactDeletes(au_id, n, ops_repo);
  for (const op of deleteOps) {
    tx.appendOp(au_id, op);
  }
  tx.deleteFactsByIds(au_id, factIdsToDelete);

  // =================================================================
  // 步骤 5-10：state 更新（内存计算）
  // =================================================================
  state.index_status = IndexStatus.STALE;
  state.last_scene_ending = await rollbackLastSceneEnding(au_id, n, ops_repo, chapter_repo);
  state.characters_last_seen = await rollbackCharactersLastSeen(
    au_id,
    n,
    ops_repo,
    chapter_repo,
    cast_registry,
    character_aliases,
  );
  state.chapter_focus = [];
  state.last_confirmed_chapter_focus = await rollbackConfirmedFocus(au_id, n, chapter_repo);

  const dirtyIdx = state.chapters_dirty.indexOf(n);
  if (dirtyIdx >= 0) {
    state.chapters_dirty.splice(dirtyIdx, 1);
  }
  delete state.chapter_titles[n];
  state.current_chapter = n;

  // undo_chapter op（主 op，包含 state snapshot 供跨设备重建）
  tx.appendOp(
    au_id,
    createOpsEntry({
      op_id: generateOpId(),
      op_type: "undo_chapter",
      target_id: chapterId,
      chapter_num: n,
      timestamp: nowUtc(),
      payload: {
        state_snapshot: {
          current_chapter: state.current_chapter,
          last_scene_ending: state.last_scene_ending,
          characters_last_seen: { ...state.characters_last_seen },
          last_confirmed_chapter_focus: [...state.last_confirmed_chapter_focus],
          chapter_titles: { ...state.chapter_titles },
          chapters_dirty: [...state.chapters_dirty],
        },
      },
    }),
  );
  tx.setState(state);

  // =================================================================
  // 事务提交：ops → chapters → facts → drafts → state
  // =================================================================
  await tx.commit(ops_repo, fact_repo, state_repo, chapter_repo, draft_repo);

  return {
    chapter_num: n,
    new_current_chapter: state.current_chapter,
  };
}

// -----------------------------------------------------------------
// 步骤 3a：facts resolves 状态回滚（收集模式，不直接写入）
// -----------------------------------------------------------------

interface ResolvesRollbackItem {
  op: OpsEntry;
  fact: Fact;
}

async function collectResolvesRollback(
  au_id: string,
  n: number,
  ops_repo: OpsRepository,
  fact_repo: FactRepository,
): Promise<ResolvesRollbackItem[]> {
  const result: ResolvesRollbackItem[] = [];

  const addFactOps = await ops_repo.get_add_facts_for_chapter(au_id, n);
  const idsToDelete = new Set(addFactOps.map((op) => op.target_id));

  if (idsToDelete.size === 0) return result;

  const allFacts = await fact_repo.list_all(au_id);

  // 找有 resolves 关系的即将被删除的 facts
  const targetsToCheck = new Set<string>();
  for (const fact of allFacts) {
    if (idsToDelete.has(fact.id) && fact.resolves) {
      targetsToCheck.add(fact.resolves);
    }
  }

  if (targetsToCheck.size === 0) return result;

  for (const targetId of targetsToCheck) {
    const target = await fact_repo.get(au_id, targetId);
    if (target === null || target.status !== FactStatus.RESOLVED) continue;

    // 检查是否有其他 fact（排除即将删除的）仍然 resolves 该目标
    const stillResolved = allFacts.some((f) => f.resolves === targetId && !idsToDelete.has(f.id) && f.id !== targetId);
    if (!stillResolved) {
      const oldStatus = target.status;
      target.status = FactStatus.UNRESOLVED;
      result.push({
        op: createOpsEntry({
          op_id: generateOpId(),
          op_type: "update_fact_status",
          target_id: targetId,
          chapter_num: n,
          timestamp: nowUtc(),
          payload: {
            old_status: oldStatus,
            new_status: FactStatus.UNRESOLVED,
            reason: UNDO_RESOLVES_CASCADE_REASON,
          },
        }),
        fact: target,
      });
    }
  }

  return result;
}

// -----------------------------------------------------------------
// 步骤 3b：回放 update_fact_status（收集模式，不直接写入）
// -----------------------------------------------------------------

async function collectManualStatusRollback(
  au_id: string,
  n: number,
  ops_repo: OpsRepository,
  fact_repo: FactRepository,
): Promise<ResolvesRollbackItem[]> {
  const result: ResolvesRollbackItem[] = [];

  const allOps = await ops_repo.list_all(au_id);
  const statusOps = allOps.filter(
    (op) =>
      op.op_type === "update_fact_status" &&
      op.chapter_num === n &&
      // 排除 undo 自身记账产生的 update_fact_status op（undo_resolves_cascade /
      // undo_manual_rollback）。它们不是本章「真实发生过」的状态变更，回放会把上一次
      // undo 的反向操作再反一次 → 二次 undo 歧义（TD-003 修复要求）。单次 undo 中这些 op
      // 还在事务里未落盘，本过滤为空操作；只在「章节 confirm→undo→reconfirm→undo」时生效。
      !isUndoGeneratedStatusOp(op),
  );

  if (statusOps.length === 0) return result;

  // 逆序回放：按 lamport_clock 降序（与 ops_projection.deterministicSort 的升序镜像，
  // timestamp / op_id 作 tiebreaker）。**不能只按 timestamp** —— nowUtc() 截到整秒
  // （file_utils.ts），同一秒内对同一 fact 的多次状态变更 timestamp 相同、比较器返回 0、
  // 排序退化为升序（正放），导致只回滚到「最后一次变更的 old_status」而非本章真正的章前态。
  // lamport_clock 在 append 时单调分配，对同秒 op 也严格有序。回放的最早一条 op（携带章前
  // old_status）最后被 push、其 tx.updateFact 最后生效（last-write-win），故须降序。
  // 前提：所有 status op 都经 file_ops.append 拿到真实 lamport_clock。导入路径（如 bundle）
  // 直接搬运 ops.jsonl 也保留原 lamport，故成立；唯有「lamport 缺失记 0」的远古遗留 op 同秒
  // 同 fact 并存时才会退到 op_id tiebreak（确定但非因果序）—— 当前数据形态不存在，见 TECH-DEBT。
  statusOps.sort((a, b) => {
    const clockA = a.lamport_clock ?? 0;
    const clockB = b.lamport_clock ?? 0;
    if (clockA !== clockB) return clockB - clockA; // lamport 降序
    if (a.timestamp !== b.timestamp) return a.timestamp < b.timestamp ? 1 : -1;
    return a.op_id < b.op_id ? 1 : a.op_id > b.op_id ? -1 : 0;
  });

  for (const op of statusOps) {
    const oldStatus = op.payload.old_status as string | undefined;
    if (!oldStatus) continue;

    const fact = await fact_repo.get(au_id, op.target_id);
    if (fact === null) continue;

    // 反向 op：把 fact 从当前状态退回 oldStatus，并落一条审计 op，使
    // rebuildFactsFromOps 重建结果与 repo 一致（TD-003 —— 此前只改 repo 不落 op，重建发散）。
    const currentStatus = fact.status;
    fact.status = oldStatus as FactStatus;
    result.push({
      op: createOpsEntry({
        op_id: generateOpId(),
        op_type: "update_fact_status",
        target_id: op.target_id,
        chapter_num: n,
        timestamp: nowUtc(),
        payload: {
          old_status: currentStatus,
          new_status: oldStatus,
          reason: UNDO_MANUAL_ROLLBACK_REASON,
        },
      }),
      fact,
    });
  }

  return result;
}

/** 该 update_fact_status op 是否为 undo 流程自身记账产生（不应被后续 undo 再反向）。 */
function isUndoGeneratedStatusOp(op: OpsEntry): boolean {
  const reason = op.payload.reason;
  return typeof reason === "string" && reason.startsWith(UNDO_REASON_PREFIX);
}

// -----------------------------------------------------------------
// 步骤 4：facts 物理删除（收集模式，不直接写入）
// -----------------------------------------------------------------

interface ChapterFactDeletes {
  deleteOps: OpsEntry[];
  factIdsToDelete: string[];
}

async function collectChapterFactDeletes(
  au_id: string,
  n: number,
  ops_repo: OpsRepository,
): Promise<ChapterFactDeletes> {
  const addFactOps = await ops_repo.get_add_facts_for_chapter(au_id, n);
  if (addFactOps.length === 0) return { deleteOps: [], factIdsToDelete: [] };

  const factIdsToDelete = addFactOps.map((op) => op.target_id);
  const deleteOps: OpsEntry[] = [];

  for (const factId of factIdsToDelete) {
    deleteOps.push(
      createOpsEntry({
        op_id: generateOpId(),
        op_type: "delete_fact",
        target_id: factId,
        chapter_num: n,
        timestamp: nowUtc(),
        payload: { reason: "undo_chapter" },
      }),
    );
  }

  return { deleteOps, factIdsToDelete };
}

// -----------------------------------------------------------------
// 步骤 6：last_scene_ending 回滚
// -----------------------------------------------------------------

async function rollbackLastSceneEnding(
  au_id: string,
  n: number,
  ops_repo: OpsRepository,
  chapter_repo: ChapterRepository,
): Promise<string> {
  if (n === 1) return "";

  // 优先：ops 快照
  const confirmOp = await ops_repo.get_confirm_for_chapter(au_id, n - 1);
  if (confirmOp) {
    const snapshot = confirmOp.payload.last_scene_ending_snapshot;
    if (typeof snapshot === "string") return snapshot;
  }

  // 降级：读取 ch{N-1} 末尾
  try {
    const content = await chapter_repo.get_content_only(au_id, n - 1);
    return extractLastSceneEnding(content);
  } catch {
    return "";
  }
}

// -----------------------------------------------------------------
// 步骤 7：characters_last_seen 回滚
// -----------------------------------------------------------------

async function rollbackCharactersLastSeen(
  au_id: string,
  n: number,
  ops_repo: OpsRepository,
  chapter_repo: ChapterRepository,
  cast_registry: { characters?: string[] },
  character_aliases: Record<string, string[]> | null,
): Promise<Record<string, number>> {
  if (n === 1) return {};

  // 优先：ops 快照
  const confirmOp = await ops_repo.get_confirm_for_chapter(au_id, n - 1);
  if (confirmOp) {
    const snapshot = confirmOp.payload.characters_last_seen_snapshot;
    if (snapshot && typeof snapshot === "object") {
      try {
        const result: Record<string, number> = {};
        for (const [k, v] of Object.entries(snapshot as Record<string, unknown>)) {
          const num = Number(v);
          // 裸 Error 有意：「快照损坏→降级扫描」的局部控制流信号，下方 catch 立即接住，
          // 从不逃逸 doUndo，故不用 UndoChapterError（逃逸域错误才用自定义类）。
          if (Number.isNaN(num)) throw new Error("invalid snapshot value");
          result[String(k)] = num;
        }
        return result;
      } catch {
        // 快照数据损坏，降级到全量扫描
      }
    }
  }

  // 降级：全量扫描重建
  return rebuildCharactersLastSeen(au_id, n, chapter_repo, cast_registry, character_aliases);
}

async function rebuildCharactersLastSeen(
  au_id: string,
  n: number,
  chapter_repo: ChapterRepository,
  cast_registry: { characters?: string[] },
  character_aliases: Record<string, string[]> | null,
): Promise<Record<string, number>> {
  // 只扫 < n 的章：第 n 章正在被撤销、此刻仍在盘上（tx 尚未 commit），若计入会把角色
  // 持久记为「最后见于已被删除的第 n 章」。姊妹路径 dirty_resolve.scanRecentChapters 同样
  // 限定 <= n-1，此处对齐（盲审 R5 正确性 M1；触发面=前一章无 confirm 快照的导入作品）。
  const chapters = await chapter_repo.list_main(au_id);
  const result: Record<string, number> = {};
  for (const ch of chapters) {
    if (ch.chapter_num >= n) continue;
    const scanned = scanCharactersInChapter(ch.content, cast_registry, character_aliases, ch.chapter_num);
    mergeCharactersLastSeen(result, scanned);
  }
  return result;
}

// -----------------------------------------------------------------
// 步骤 9：last_confirmed_chapter_focus 回退
// -----------------------------------------------------------------

async function rollbackConfirmedFocus(au_id: string, n: number, chapter_repo: ChapterRepository): Promise<string[]> {
  if (n <= 1) return [];
  try {
    const prevCh = await chapter_repo.get(au_id, n - 1);
    return prevCh ? [...prevCh.confirmed_focus] : [];
  } catch (err) {
    // best-effort：focus 回滚不值得让 undo 级联失败，但读错误落日志可诊断
    logCatch("undo", `rollbackConfirmedFocus read failed for ch${n - 1}`, err);
    return [];
  }
}
